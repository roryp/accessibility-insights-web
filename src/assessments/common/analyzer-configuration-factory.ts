// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Messages } from 'common/messages';
import { RuleAnalyzerConfiguration } from 'injected/analyzers/analyzer';
import { ScannerUtils } from 'injected/scanner-utils';

export class AnalyzerConfigurationFactory {
    public static forScanner(base: Partial<RuleAnalyzerConfiguration>): RuleAnalyzerConfiguration {
        const defaultValues: Partial<RuleAnalyzerConfiguration> = {
            resultProcessor: (scanner: ScannerUtils) => scanner.getAllCompletedInstances,
            telemetryProcessor: telemetryFactory => telemetryFactory.forAssessmentRequirementScan,
            analyzerMessageType: Messages.Assessment.AssessmentScanCompleted,
        };

        return {
            ...defaultValues,
            ...base,
        } as RuleAnalyzerConfiguration;
    }
}
