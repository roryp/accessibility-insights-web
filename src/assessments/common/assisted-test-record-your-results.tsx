// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import * as Markup from '../markup';

export const AssistedTestRecordYourResults = () => (
    <li>
        Record your results:
        <ol>
            <li>
                Select <Markup.Term>Fail</Markup.Term> for any instances that do not meet the
                requirement.
            </li>
            <li>
                Otherwise, select <Markup.Term>Pass</Markup.Term>. Or, after you have marked all
                failures, select <Markup.Term>Pass unmarked instances.</Markup.Term>
            </li>
        </ol>
    </li>
);
