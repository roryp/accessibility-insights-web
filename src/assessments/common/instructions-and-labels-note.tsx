// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from 'react';

import * as Markup from '../markup';

export const InstructionsAndLabelsNotes = () => (
    <Markup.Emphasis>
        Note: For WCAG 2.1 compliance, both visible instructions and labels must be programmatically
        related to a widget. For WCAG 2.0 compliance, only visible instructions must be
        programmatically related to a widget.
    </Markup.Emphasis>
);
