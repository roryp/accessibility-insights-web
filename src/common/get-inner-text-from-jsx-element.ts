// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
export function getInnerTextFromJsxElement(element: JSX.Element): string {
    return getInnerTextFromJsxElementRecursive(element).replace(/\s+/g, ' ').trim();
}

function getInnerTextFromJsxElementRecursive(element: JSX.Element): string {
    if (typeof element !== 'object') {
        return `${element}`;
    }

    const childProps = element.props.children || [];
    if (Array.isArray(childProps)) {
        return childProps.map(child => getInnerTextFromJsxElementRecursive(child)).join('');
    }

    return `${childProps}`;
}
