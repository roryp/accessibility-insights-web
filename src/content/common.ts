// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as react from 'react';

import { ContentCreator } from 'views/content/content-page';
import { GuidanceTitle as gt } from 'views/content/guidance-title';
import { link } from './link';

export const create = ContentCreator(link);
export const React = react;
export const GuidanceTitle = gt;
