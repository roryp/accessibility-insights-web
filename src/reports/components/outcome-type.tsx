// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { CheckIcon, CheckIconInverted } from 'common/icons/check-icon';
import { CircleIcon } from 'common/icons/circle-icon';
import { CrossIcon, CrossIconInverted } from 'common/icons/cross-icon';
import { InapplicableIcon, InapplicableIconInverted } from 'common/icons/inapplicable-icon';
import * as React from 'react';

import { InstanceOutcomeType } from './instance-outcome-type';
import { RequirementOutcomeType } from './requirement-outcome-type';

export type OutcomeUnits = 'percentage' | 'requirements';

export type OutcomeStats = { [OT in OutcomeType]: number };
export type OutcomeType = RequirementOutcomeType | InstanceOutcomeType;

export interface OutcomeTypeSemantic {
    pastTense: string;
}

export const outcomeTypeSemantics: { [OT in OutcomeType]: OutcomeTypeSemantic } = {
    pass: { pastTense: 'Passed' },
    incomplete: { pastTense: 'Incomplete' },
    fail: { pastTense: 'Failed' },
    inapplicable: { pastTense: 'Not applicable' },
    review: { pastTense: 'Needs review' },
};

export const outcomeIconMap: { [OT in OutcomeType]: JSX.Element } = {
    pass: <CheckIcon />,
    incomplete: <CircleIcon />,
    fail: <CrossIcon />,
    inapplicable: <InapplicableIcon />,
    review: <CircleIcon />,
};

export const outcomeIconMapInverted: { [OT in OutcomeType]: JSX.Element } = {
    pass: <CheckIconInverted />,
    incomplete: <CircleIcon />,
    fail: <CrossIconInverted />,
    inapplicable: <InapplicableIconInverted />,
    review: <CircleIcon />,
};
