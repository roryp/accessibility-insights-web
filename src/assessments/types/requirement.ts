// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { UniquelyIdentifiableInstances } from 'background/instance-identifier-generator';
import { ManualTestStatus } from 'common/types/manual-test-status';
import {
    AssessmentNavState,
    GeneratedAssessmentInstance,
    InstanceIdToInstanceDataMap,
} from 'common/types/store-data/assessment-result-data';
import { FeatureFlagStoreData } from 'common/types/store-data/feature-flag-store-data';
import { DetailsViewActionMessageCreator } from 'DetailsView/actions/details-view-action-message-creator';
import {
    AssessmentInstanceRowData,
    AssessmentInstanceTable,
} from 'DetailsView/components/assessment-instance-table';
import { RequirementLink } from 'DetailsView/components/requirement-link';
import { Analyzer } from 'injected/analyzers/analyzer';
import { AnalyzerProvider } from 'injected/analyzers/analyzer-provider';
import { DecoratedAxeNodeResult } from 'injected/scanner-utils';
import {
    PropertyBags,
    VisualizationInstanceProcessorCallback,
} from 'injected/visualization-instance-processor';
import { Drawer } from 'injected/visualization/drawer';
import { DrawerProvider } from 'injected/visualization/drawer-provider';
import { IColumn } from 'office-ui-fabric-react';
import { DictionaryStringTo } from 'types/common-types';
import { ContentPageComponent, HyperlinkDefinition } from 'views/content/content-page';
import { IGetMessageGenerator } from '../assessment-default-message-generator';
import { InstanceTableColumn } from './instance-table-column';
import { ReportInstanceFields } from './report-instance-field';

export interface Requirement {
    key: string;
    description: JSX.Element;
    order?: number;
    name: string;
    renderStaticContent?: () => JSX.Element;
    howToTest: JSX.Element;
    addFailureInstruction?: string;
    infoAndExamples?: ContentPageComponent;
    isManual: boolean;
    // This is for semi-manual cases where we can't present a list of instances like an assisted
    // test would, but can infer a PASS or FAIL state. If not specified, acts like () => UNKNOWN.
    getInitialManualTestStatus?: (instances: InstanceIdToInstanceDataMap) => ManualTestStatus;
    guidanceLinks: HyperlinkDefinition[];
    columnsConfig?: InstanceTableColumn[];
    getAnalyzer?: (provider: AnalyzerProvider) => Analyzer;
    getVisualHelperToggle?: (props: VisualHelperToggleConfig) => JSX.Element;
    // Any results this returns false for will be omitted from visual helper displays, but still
    // present for the purposes of instance lists or getInitialManualTestStatus. By default, all
    // results support visualization.
    isVisualizationSupportedForResult?: (result: DecoratedAxeNodeResult) => boolean;
    visualizationInstanceProcessor?: VisualizationInstanceProcessorCallback<
        PropertyBags,
        PropertyBags
    >;
    getDrawer?: (provider: DrawerProvider, featureFlagStoreData?: FeatureFlagStoreData) => Drawer;
    getNotificationMessage?: (selectorMap: DictionaryStringTo<any>) => string;
    doNotScanByDefault?: boolean;
    switchToTargetTabOnScan?: boolean;
    generateInstanceIdentifier?: (instance: UniquelyIdentifiableInstances) => string;
    reportInstanceFields?: ReportInstanceFields;
    renderReportDescription?: () => JSX.Element;
    getInstanceStatus?: (result: DecoratedAxeNodeResult) => ManualTestStatus;
    getInstanceStatusColumns?: () => Readonly<IColumn>[];
    getDefaultMessage?: IGetMessageGenerator;
    renderInstanceTableHeader?: (
        table: AssessmentInstanceTable,
        items: AssessmentInstanceRowData[],
    ) => JSX.Element;
    renderRequirementDescription?: (requirementLink: RequirementLink) => JSX.Element;
}

export type VisualHelperToggleConfigDeps = {
    detailsViewActionMessageCreator: DetailsViewActionMessageCreator;
};

export interface VisualHelperToggleConfig {
    deps: VisualHelperToggleConfigDeps;
    assessmentNavState: AssessmentNavState;
    instancesMap: DictionaryStringTo<GeneratedAssessmentInstance>;
    isStepEnabled: boolean;
    isStepScanned: boolean;
}
