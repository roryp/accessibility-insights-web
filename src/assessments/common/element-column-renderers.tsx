// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { InstanceTableRow } from 'assessments/types/instance-table-data';
import { AssessmentInstanceDetailsColumn } from 'DetailsView/components/assessment-instance-details-column';
import * as React from 'react';

export function onRenderPathColumn(item: InstanceTableRow): JSX.Element {
    let textContent = '';
    if (item.instance.target) {
        textContent = item.instance.target.join(';');
    }

    return (
        <AssessmentInstanceDetailsColumn
            background={null}
            labelText={null}
            textContent={textContent}
            tooltipId={null}
            customClassName="not-applicable"
        />
    );
}

export function onRenderSnippetColumn(item: InstanceTableRow): JSX.Element {
    return (
        <AssessmentInstanceDetailsColumn
            background={null}
            labelText={null}
            textContent={item.instance.html}
            tooltipId={null}
            customClassName="not-applicable"
        />
    );
}
