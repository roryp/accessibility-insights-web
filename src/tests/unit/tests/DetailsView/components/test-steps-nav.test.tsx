// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { INavLink } from 'office-ui-fabric-react';
import * as React from 'react';
import * as TestUtils from 'react-dom/test-utils';
import { Mock, Times } from 'typemoq';

import { AssessmentsProvider } from 'assessments/types/assessments-provider';
import { Requirement } from 'assessments/types/requirement';
import { OutcomeTypeSemantic } from 'reports/components/outcome-type';
import { outcomeTypeSemanticsFromTestStatus } from 'reports/components/requirement-outcome-type';
import { getInnerTextFromJsxElement } from '../../../../../common/get-inner-text-from-jsx-element';
import { ManualTestStatus } from '../../../../../common/types/manual-test-status';
import { DetailsViewActionMessageCreator } from '../../../../../DetailsView/actions/details-view-action-message-creator';
import {
    TestStepNavProps,
    TestStepsNav,
} from '../../../../../DetailsView/components/test-steps-nav';
import { EventStubFactory } from '../../../common/event-stub-factory';
import {
    CreateTestAssessmentProvider,
    CreateTestAssessmentProviderAutomated,
} from '../../../common/test-assessment-provider';

class TestableTestStepsNav extends TestStepsNav {
    public getOnTestStepSelected(): (
        event?: React.MouseEvent<HTMLElement>,
        item?: INavLink,
    ) => void {
        return this.onTestStepSelected;
    }

    public getRenderNavLink(): (link: INavLink) => JSX.Element {
        return this.renderNavLink;
    }
}

describe('TestStepsNav', () => {
    it('renders assisted tests', () => {
        runTest(CreateTestAssessmentProvider());
    });

    it('renders automated tests', () => {
        runTest(CreateTestAssessmentProviderAutomated());
    });

    function runTest(assessmentProvider: AssessmentsProvider): void {
        const eventFactory = new EventStubFactory();
        const detailsViewActionMessageCreatorMock = Mock.ofType(DetailsViewActionMessageCreator);
        const eventStub = eventFactory.createKeypressEvent() as any;

        const all = assessmentProvider.all();
        const assessment = all[0];
        const firstStep = assessment.requirements[0];

        const item = {
            key: firstStep.key,
            name: firstStep.name,
            description: firstStep.description,
            url: '',
            index: 1,
            forceAnchor: true,
            renderRequirementDescription: firstStep.renderRequirementDescription,
        };

        const props: TestStepNavProps = {
            deps: {
                detailsViewActionMessageCreator: detailsViewActionMessageCreatorMock.object,
                assessmentsProvider: assessmentProvider,
                getInnerTextFromJsxElement: getInnerTextFromJsxElementStub(),
                outcomeTypeSemanticsFromTestStatus: createOutcomeTypeSemanticsFromTestStatusStub(),
            },
            selectedTest: assessment.visualizationType,
            selectedTestStep: firstStep.key,
            stepStatus: {},
            assessmentsProvider: assessmentProvider,
            ariaLabel: 'test',
        };

        generateStepStatus(assessment.requirements, props);

        detailsViewActionMessageCreatorMock
            .setup(a => a.selectRequirement(eventStub, item.key, props.selectedTest))
            .verifiable(Times.once());

        const component = React.createElement(TestableTestStepsNav, props);
        const testObject = TestUtils.renderIntoDocument(component);
        const rendered = testObject.render();
        rendered.props.onLinkClick(eventStub, item);

        expect(rendered).toMatchSnapshot('render');
        expect(testObject.getRenderNavLink()(item)).toMatchSnapshot('getRenderNavLink');
        detailsViewActionMessageCreatorMock.verifyAll();
    }

    function generateStepStatus(testSteps: Requirement[], props: TestStepNavProps): void {
        testSteps.forEach((step, index) => {
            props.stepStatus[step.key] = {
                stepFinalResult: index % 2 === 0 ? ManualTestStatus.UNKNOWN : ManualTestStatus.PASS,
                isStepScanned: false,
                name: step.name,
            };
        });
    }
    function createOutcomeTypeSemanticsFromTestStatusStub(): typeof outcomeTypeSemanticsFromTestStatus {
        const outcomeTypeSemanticsFromTestStatusMock = Mock.ofInstance(
            outcomeTypeSemanticsFromTestStatus,
        );

        outcomeTypeSemanticsFromTestStatusMock
            .setup(f => f(ManualTestStatus.PASS))
            .returns(() => {
                return { pastTense: 'passed' } as OutcomeTypeSemantic;
            });
        outcomeTypeSemanticsFromTestStatusMock
            .setup(f => f(ManualTestStatus.UNKNOWN))
            .returns(() => {
                return { pastTense: 'unknown' } as OutcomeTypeSemantic;
            });

        return outcomeTypeSemanticsFromTestStatusMock.object;
    }

    function getInnerTextFromJsxElementStub(): typeof getInnerTextFromJsxElement {
        return status => {
            return 'some test step description';
        };
    }
});
