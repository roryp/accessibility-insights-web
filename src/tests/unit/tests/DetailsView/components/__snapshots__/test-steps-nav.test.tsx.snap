// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestStepsNav renders assisted tests: getRenderNavLink 1`] = `
<RequirementLink
  link={
    Object {
      "description": <div>
         assessment-1-step-1 description
      </div>,
      "forceAnchor": true,
      "index": 1,
      "key": "assessment-1-step-1",
      "name": "assessment-1-step-1-name",
      "renderRequirementDescription": [Function],
      "url": "",
    }
  }
  renderRequirementDescription={[Function]}
  status={1}
/>
`;

exports[`TestStepsNav renders assisted tests: render 1`] = `
<StyledNavBase
  ariaLabel="test"
  className="test-step-nav"
  groups={
    Array [
      Object {
        "links": Array [
          Object {
            "description": <div>
               assessment-1-step-1 description
            </div>,
            "forceAnchor": true,
            "index": 1,
            "key": "assessment-1-step-1",
            "name": "assessment-1-step-1-name",
            "renderRequirementDescription": [Function],
            "title": "assessment-1-step-1-name. unknown. some test step description",
            "url": "",
          },
          Object {
            "description": <div>
               assessment-1-step-2 description
            </div>,
            "forceAnchor": true,
            "index": 2,
            "key": "assessment-1-step-2",
            "name": "assessment-1-step-2-name",
            "renderRequirementDescription": [Function],
            "title": "assessment-1-step-2-name. passed. some test step description",
            "url": "",
          },
          Object {
            "description": <div>
               assessment-1-step-3 description
            </div>,
            "forceAnchor": true,
            "index": 3,
            "key": "assessment-1-step-3",
            "name": "assessment-1-step-3-name",
            "renderRequirementDescription": [Function],
            "title": "assessment-1-step-3-name. unknown. some test step description",
            "url": "",
          },
        ],
      },
    ]
  }
  onLinkClick={[Function]}
  onRenderLink={[Function]}
  selectedKey="assessment-1-step-1"
/>
`;

exports[`TestStepsNav renders automated tests: getRenderNavLink 1`] = `
<RequirementLink
  link={
    Object {
      "description": null,
      "forceAnchor": true,
      "index": 1,
      "key": "assessment-3-step-2",
      "name": "assessment-3-step-2-name",
      "renderRequirementDescription": [Function],
      "url": "",
    }
  }
  renderRequirementDescription={[Function]}
  status={1}
/>
`;

exports[`TestStepsNav renders automated tests: render 1`] = `
<StyledNavBase
  ariaLabel="test"
  className="test-step-nav"
  groups={
    Array [
      Object {
        "links": Array [
          Object {
            "description": null,
            "forceAnchor": true,
            "index": 1,
            "key": "assessment-3-step-2",
            "name": "assessment-3-step-2-name",
            "renderRequirementDescription": [Function],
            "title": "assessment-3-step-2-name. unknown. some test step description",
            "url": "",
          },
          Object {
            "description": null,
            "forceAnchor": true,
            "index": 2,
            "key": "assessment-3-step-1",
            "name": "assessment-3-step-1-name",
            "renderRequirementDescription": [Function],
            "title": "assessment-3-step-1-name. passed. some test step description",
            "url": "",
          },
        ],
      },
    ]
  }
  onLinkClick={[Function]}
  onRenderLink={[Function]}
  selectedKey="assessment-3-step-2"
/>
`;
